### ISSUE TITLE
Good metrics but model still fails ! Help needed pls 🙏

### ISSUE THREAD

**Issue by DavidScanu** [ultralytics/ultralytics#10116]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

As a final project for my diploma, I am training a custom Yolo model for detecting **boats from aerial views**.

## Dataset 

The dataset is composed of images extracted from 23 YouTube videos where I extracted 1 image every 30 frames. It has **4 classes** (boat, cruise shp, sailboat, cargo) and is split like this : 
- Train set : 1290 images (430 x 3 with data augmentation)
- Valid set : 118 images 
- Test set : 61 images 

The dataset is available here : https://universe.roboflow.com/david-scanu-mfhtq/aerial-views-of-boats/dataset/1

## Results from multiple runs

Training notebook : https://colab.research.google.com/drive/1R9Rb8FRNEGeHdfeHdDyOKfAEbkOrSPEB?usp=sharing

![metrics-yolo-boats](https://github.com/ultralytics/ultralytics/assets/7550631/ac8a781d-455b-4fe7-8e95-462605e3a6ed)

## Metrics for run-07 (yolov8n.pt)

Evaluation notebook : https://colab.research.google.com/drive/1vzwMdKhunRnvfjVaVUT91pj-GMLQyNVO?usp=sharing

### Training results 

![results](https://github.com/ultralytics/ultralytics/assets/7550631/2251c2ea-d34c-467a-9439-fef5b49872cc)

### Confusion Matrix

![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/7550631/d3c87803-2908-483d-a83f-0df328cf1a67)

### Metrics on validation set 

```
metrics = custom_model.val(plots=True)
```

![metrics-run-07-val](https://github.com/ultralytics/ultralytics/assets/7550631/b1f8a9a6-ea0a-45b9-848b-ea5c045b5af5)

### Metrics on test set

```
metrics = custom_model.val(split="test", plots=True)
```

![metrics-run-07-test](https://github.com/ultralytics/ultralytics/assets/7550631/a96d4c52-c4ef-4668-bca1-a88dc937d1f9)

## Here is the weird part !!! 

When I try my model on new random images from the web. **My model fail to detect obvious boats** or mistaken some classes. 

![detections-weird](https://github.com/ultralytics/ultralytics/assets/7550631/e7f4b144-c53c-444b-9db7-83df44af8832)

## Questions

1. **Why is my model failing to detect on random unseen images ?** 
2. Why are my results from different runs so close together ? (despite them using different yolo pre-trained weights)
3. Is it wrong to have too many images that are similar because they come from the same videos ? 

Any help or advices woud be greatly appreciated ! My final exam/presentation is in about a month and I really want to pass.

**Comment by glenn-jocher** [comment]
@DavidScanu hey there! 🌊

It looks like you've put in a lot of effort into your project, and your thorough approach and documentation are commendable! Let's dive into the issues you're facing.

1. **Detection Failures on Random Images:** It's possible that the new images you're testing with have variations (angle, lighting, background) not well-represented in your training set. Consider expanding your dataset with more diverse images that cover more variations.

2. **Consistent Results Across Different Runs:** If the runs are fairly similar in terms of dataset and initialization weights, the results might not vary significantly. This consistency isn't necessarily bad, as it suggests stability in your training process.

3. **Similar Images in Dataset:** Using frames from videos can introduce high correlation between training images, which might not generalize well. Try adding more unique images to your dataset or employing techniques like **increased data augmentation** to introduce more variety.

For improving your model’s performance on unseen images, consider these tips:

- **Diverse Dataset**: Ensure your dataset is diverse. Collect images from various sources that represent different conditions.
  
- **Augmentation**: Increase data augmentation to expose your model to a wide range of variations.

- **Fine-tuning**: If possible, fine-tune your model with a small set of images that represent the kind of images you expect to encounter in real-world applications.

- **Analyze Failures**: Investigate the common characteristics of misclassified or undetected objects. It can provide insights into what your model is missing.

Here's a small snippet on how you can increase augmentation for your training:

```python
# Assuming you're using ultralytics YOLO, add to your YAML file
augment:
  perspective: 0.0  # up to 0.1
  translate: 0.1  # up to 0.5 for more translation
  scale: 0.5  # up to 2.0 for more scaling
  shear: 0.0  # up to 0.1
  flipud: 0.0  # up to 0.1
  fliplr: 0.5  # up to 0.5
  mosaic: 1.0  # ensure mosaic is enabled
  mixup: 0.0  # up to 0.1
```

Lastly, keep iterating! 🔄 Model improvement is a continuous process and involves a lot of experimentation. You're on the right track, and with some adjustments, you should see better results. Good luck with your final presentation! You got this! 🚀

**Comment by fiksii-copilot** [comment]
Moreover, your recall is just 90%. So, you will be missing 10% of objects

**Comment by DavidScanu** [comment]
@glenn-jocher Thanks a lot for your elaborate response ! Thanks a lot for all these insights. I will definitly be able to elborate explanations and improvement clues during my presentation (this is what matters the most). 

I was supecting my dataset needed more diversity. So my goal would be **adding fresh new images from various sources**. 

### Class imbalance

![chart](https://github.com/ultralytics/ultralytics/assets/7550631/f1f049e4-b43e-49a2-9fdd-cfef31edfc60)

1. Results for the class "boat" are lower than other classes. Is this due to this class being way more present than others ? 
2. Does having a small val and test set impact metric evaluation ?
3. **Data augmentation** : Should I be adding the code snippet you provided to my **data.yaml** ? Do I need to add `augment=True` to `model.train()` ? (if you have a notebook example, ti would be great). 
4. Does Yolo perform augmentation as a default under the hood ? (I think i saw this somewhere in the doc)

Thanks again for taking the time to answer. I greatly appreciate it.

**Comment by DavidScanu** [comment]
@nhjydywd Thanks a lot for your answer. I agree on all those points. I will try to add more images to my dataset.
